internal/pkg/instrumentation/probe/sampling/sampling.go

```diff
@@ -0,0 +1,305 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package sampling
+
+import (
+	"bytes"
+	"encoding"
+	"encoding/binary"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/cilium/ebpf"
+)
+
+const (
+	samplersConfigMapName     = "samplers_config_map"
+	probeActiveSamplerMapName = "probe_active_sampler_map"
+)
+
+// Config is used to configure the samplers used by eBPF.
+type Config struct {
+	samplersConfig     *ebpf.Map
+	probeActiveSampler *ebpf.Map
+
+	currentSamplerID samplerID
+}
+
+type samplerType uint64
+
+const (
+	// OpenTelemetry spec-defined samplers.
+	samplerAlwaysOn samplerType = iota
+	samplerAlwaysOff
+	samplerTraceIDRatio
+	samplerParentBased
+	// Custom samplers.
+)
+
+type traceIDRatioConfig struct {
+	// samplingRateNumerator is the numerator of the sampling rate.
+	// see samplingRateDenominator for more information.
+	samplingRateNumerator uint64
+}
+
+// samplerID is a unique identifier for a sampler. It is used as a key in the samplers config map,
+// and as a value in the active sampler map. In addition samplers can reference other samplers in their configuration by their ID.
+type samplerID uint32
+
+type parentBasedConfig struct {
+	Root             samplerID
+	RemoteSampled    samplerID
+	RemoteNotSampled samplerID
+	LocalSampled     samplerID
+	LocalNotSampled  samplerID
+	_                [4]byte
+}
+
+// the following are constants which are used by the eBPF code.
+// they should be kept in sync with the definitions there.
+const (
+	maxSampleConfigDataSize = 256
+	sampleConfigSize        = maxSampleConfigDataSize + 8
+	// since eBPF does not support floating point arithmetic, we use a rational number to represent the ratio.
+	// the denominator is fixed and the numerator is used to represent the ratio.
+	// This value can limit the precision of the sampling rate, hence setting it to a high value should be enough in terms of precision.
+	samplingRateDenominator = 1e9
+	maxSamplers             = 32
+)
+
+const (
+	tracesSamplerKey    = "OTEL_TRACES_SAMPLER"
+	tracesSamplerArgKey = "OTEL_TRACES_SAMPLER_ARG"
+
+	samplerNameAlwaysOn                = "always_on"
+	samplerNameAlwaysOff               = "always_off"
+	samplerNameTraceIDRatio            = "traceidratio"
+	samplerNameParentBasedAlwaysOn     = "parentbased_always_on"
+	samplerNameParsedBasedAlwaysOff    = "parentbased_always_off"
+	samplerNameParentBasedTraceIDRatio = "parentbased_traceidratio"
+)
+
+// config data for samplers is a union of all possible sampler configurations.
+// the size of the data is fixed, and the actual configuration is stored in the first part of the data.
+// the rest of the data is padding to make sure the size is fixed.
+
+// The spec-defined samplers have a constant ID, and are always available.
+const (
+	alwaysOnID     samplerID = 0
+	alwaysOffID    samplerID = 1
+	traceIDRatioID samplerID = 2
+	parentBasedID  samplerID = 3
+)
```

Comment:
These look like duplicates of https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/971/files#diff-066535099a2e1f5733afadd0071ea145869090928742aa0b51daf7a764c54698R33-R40. Why define these twice?